### ISSUE TITLE
Feature Request: Add sFID as a Metric

### ISSUE THREAD

**Issue by lavinal712** [chaofengc/IQA-PyTorch#259]
tensorflow implement: https://github.com/[openai/guided-diffusion](https://github.com/openai/guided-diffusion)

Request for pytorch implement.

**Comment by chaofengc** [comment]
`sFID` is now supported and can be used via create_metric('sfid').

Please note that the computed values may differ slightly from those in https://github.com/openai/guided-diffusion/blob/main/evaluations/evaluator.py, due to potential variations in the Inception network weights and other minor numerical differences.

If you’re able to share any results or comparisons, that would be very helpful for better understanding and documenting these differences. Thanks!

**Comment by lavinal712** [comment]
I tested the FID and sFID using the reference files provided by `guided_diffusion`, but the results did not meet expectations.

I cloned IQA-PyTorch locally and made the following modifications to `fid_arch.py`:

```python
def get_folder_features(
    fdir,
    model=None,
    num_workers=12,
    batch_size=32,
    test_img_size=(299, 299),
    device=torch.device('cuda'),
    mode='clean',
    description='',
    verbose=True,
):
    r"""
    Compute the inception features for a folder of image files
    """
    # files = scandir_images(fdir)

    # if verbose:
    #     print(f'Found {len(files)} images in the folder {fdir}')

    # dataset = ResizeDataset(files, mode=mode, size=test_img_size)

    images = np.load(fdir)['arr_0']

    if verbose:
        print(f'Found {len(images)} images in the file {fdir}')

    dataset = ResizeDataset(images, mode=mode, size=test_img_size)

    dataloader = torch.utils.data.DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=False,
        drop_last=False,
        num_workers=num_workers,
    )
...

class ResizeDataset(torch.utils.data.Dataset):
    """
    A placeholder Dataset that enables parallelizing the resize operation
    using multiple CPU cores
    files: list of all files in the folder
    mode:
        - clean: use PIL resize before calculate features
        - legacy_pytorch: do not resize here, but before pytorch model
    """

    def __init__(self, files, mode, size=(299, 299)):
        self.files = files
        self.size = size
        self.mode = mode

    def __len__(self):
        return len(self.files)

    def __getitem__(self, i):
        # path = str(self.files[i])
        # img_pil = Image.open(path).convert('RGB')
        img_pil = Image.fromarray(self.files[i].astype(np.float32), mode='RGB')
```

Code:
```python
import pyiqa

fid_metric = pyiqa.create_metric('sfid') # pyiqa.create_metric('fid')
score = fid_metric('VIRTUAL_imagenet256_labeled.npz', 'admnet_guided_upsampled_imagenet256.npz')
print(score)
```

Output: 
```
fid: 1.9601370501265478
sfid: 0.31123380564500636
```

Expect:
```
fid: 3.9425574129223264
sfid: 6.140433703346162
```

**Comment by chaofengc** [comment]
Thank you for the information.

As mentioned in our previous response, the results are expected to differ from those obtained using `guided_diffusion`, primarily due to differences in the Inception model weights. Our tool is mainly based on the following repositories:

* https://github.com/mseitzer/pytorch-fid
* https://github.com/toshas/torch-fidelity

If you have access to the specific Inception weights used in `guided_diffusion`, we’d appreciate it if you could share that information.

**Comment by lavinal712** [comment]
The Inception in `guided_diffusion` is https://openaipublic.blob.core.windows.net/diffusion/jul-2021/ref_batches/classify_image_graph_def.pb. This is for TensorFlow, so the weight names have some difference.

https://github.com/openai/guided-diffusion/blob/main/evaluations/evaluator.py shows how to use it.

**Comment by chaofengc** [comment]
Thank you for the information. The issue isn’t only about the weight names—there are multiple versions of Inception weights available. For instance, the version used in `torchvision` differs from others, and I’m not sure whether the version used in `guided_diffusion` aligns with the original FID paper from https://github.com/bioinf-jku/TTUR or the widely used PyTorch reproduction by https://github.com/mseitzer/pytorch-fid.

While the `guided_diffusion` Inception model has been used in many papers, converting TensorFlow weights to PyTorch in a consistent and reliable way can be time-consuming. If you're able to help with that process, it would be greatly appreciated.

**Comment by lavinal712** [comment]
I will check it.

**Comment by chaofengc** [comment]
I have verified that the Inception weights used in `guided_diffusion` are the same as those from https://github.com/bioinf-jku/TTUR and https://github.com/mseitzer/pytorch-fid. 

Additionally, I’ve fixed the `sfid` metric. The results from the latest version should now closely match those from `guided_diffusion`. However, due to numerical differences and varying implementations between PyTorch and TensorFlow, there may still be slight discrepancies. A variation within ±1% is considered normal.

**Comment by lavinal712** [comment]
Thanks, I'll try different methods to verify this.

**Comment by lavinal712** [comment]
```
fid: 7.477698307564026
sfid: 22.831740039075612
```

I saved the images from the `npz` file and performed the verification. There's still a significant discrepancy compared to the results in the original paper. Not only sFID, but FID also has issues.

Is there a reproducible FID verification I can use to determine if it's an issue with my code?

**Comment by chaofengc** [comment]
As previously mentioned, our fid implementation aligns with https://github.com/mseitzer/pytorch-fid.

**Comment by lavinal712** [comment]
Thank you, solved.
